File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

fun TreeNode.hasChild(): Boolean {
    return this.left != null || this.right != null
}

class Solution {
    fun sumNumbers(root: TreeNode?): Int {
        if (root == null) return 0
        
        // 자식의 문자열 목록들을 모두 받아와서 sum에 합치고 리턴한다.
        val list = getChildStrs(root)
        return list.map { it -> it.toInt() }.sum()
    }
    
    private fun getChildStrs(root: TreeNode?): List<String> {
        if (root == null) return emptyList()
        
        if (!root.hasChild()) {
            return listOf(root.`val`.toString())
        }
        
        // 자식들의 리스트를 얻어서 합친다.
        val left = getChildStrs(root.left)
        val right = getChildStrs(root.right)
        
        return listOf(left, right).flatten()
            .map { it -> root.`val`.toString() + it }
    }
}

File: LeetCode/Medium/0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        
        if root.left is None and root.right is None:
            return root.val
        
        if root.left is None:
            return self.getTotal(root.right, root.val)
        elif root.right is None:
            return self.getTotal(root.left, root.val)
        else:
            left = self.getTotal(root.left, root.val)
            right = self.getTotal(root.right, root.val)
            return left + right
    
    def getTotal(self, root: Optional[TreeNode], prefix: int) -> int:
        if root is None:
            return prefix
        
        if root.left is None and root.right is None:
            return int(str(prefix) + str(root.val))
        
        newPrefix = int(str(prefix) + str(root.val)) 
        if root.left is None:
            return self.getTotal(root.right, newPrefix)
        elif root.right is None:
            return self.getTotal(root.left, newPrefix)
        
        return self.getTotal(root.left, newPrefix) + self.getTotal(root.right, newPrefix)
    